### PR TITLE
Vberenz/main sim

### PR DESCRIPTION
**Pull request by vincentberenz**
[//]: # "Thanks for your contribution.  To make life of the reviewers easier,"
[//]: # "please give this pull request a meaningful title and provide the"
[//]: # "requested information below."

## Description
Main addition : the MainSim class and its unit tests. The MainSim class will be used to control the robot and the ball of the "main simulation", i.e. the mujoco simulation that mirrors the robot while playing recorded ball trajectories.

Other: fixed some docstrings, fixed some minor errors, added test_combinations.py which test the pseudo-robot, main sim and extra balls sets working together, refactored a bit types.py, added reset method and unit tests to ExtraBallsSet  

## How I Tested

Unit tests

## I fulfilled the following requirements

[//]: # "Please make sure you followed these steps before requesting a review."
[//]: # "Check the boxes in the list below, when done."

- [x] All new code is formatted according to our style guide (for C++ run clang-format, for Python, run flake8 and fix all warnings).
- [x] All new functions/classes are documented and existing documentation is updated according to changes.
- [x] No commented code from testing/debugging is kept (unless there is a good reason to keep it).

### CODE REVIEW

File: hysr/main_sim.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+    def reset_contact(self) -> None:
+        """
+        Reset the contact between the ball and the racket,
+        i.e. past contacts will be 'forgotten' and the 
+        o80 backend regains control of the ball.
+        """
+        self._handle.reset_contact(_ball_segment_id)
+
+    def get(self) -> types.MainSimState:
```
Comment (luator) [inline hysr/main_sim.py:168]: Just `get` is a bit vague, maybe something like `get_state` would be better?

**Comment by dtrb** [comment]
Honestly, I'm overwhelmed by the pure number of lines as I don't exactly know what the previous code has done differently. All looks good to me. I agree with Felix' suggestions.

**Review by dtrb — APPROVED** [review]

File: hysr/extra_balls.py [hunk 2]
```diff
@@ -48,15 +35,23 @@ class ExtraBallsSet:
...
-    Args:
-         setid: id of the extra ball set (arbitrary, but must be different of all sets)
-         nb_balls: has to be 3, 10, 20, 50 or 100
-         graphics: if the mujoco simulation should run graphics
-         scene : position and orientation of the table and robot
-         contact: which contact between the enviromnent and the balls should be
-                  monitored (default: the racket of the robot)
-         trajectory_getter: instance of ball_behavior.TrajectoryGetter. Will be used
-                            to set the trajectories the balls will be required to follow
+    Arguments
```
[the remark this answers is not in the document]
Comment (vincentberenz) [reply hysr/extra_balls.py:38]: Indeed, it escaped my mind. I find that numpy style looks better. But we already have some code in google style. I see no issue in having some repo in google style and other in numpy style, but maybe I am overlooking something ?

File: hysr/main_sim.py (hunk 1 above)
Comment (vincentberenz) [reply hysr/main_sim.py]: updated !

File: hysr/extra_balls.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (vincentberenz) [reply hysr/extra_balls.py:38]: I merge for now, happy to update the doc later if we decide to enforce google style

[the remark this answers is not in the document]
Comment (luator) [reply hysr/extra_balls.py:38]: Both styles are supported by Sphinx' Napoleon plugin, so I don't see a problem with using both in parallel.

**Pull request merged**